src/transform/src/join_implementation.rs

```diff
@@ -206,118 +235,108 @@ impl JoinImplementation {
-                        })
-                    });
+                    _ => {}
                 }
+                available_arrangements[index].sort();
+                available_arrangements[index].dedup();
+                let reverse_project = project
+                    .into_iter()
+                    .enumerate()
+                    .map(|(i, c)| (c, i))
+                    .collect::<BTreeMap<_, _>>();
+                // Eliminate arrangements referring to columns that have been
+                // projected away by surrounding MFPs.
+                available_arrangements[index].retain(|key| {
+                    key.iter()
+                        .all(|k| k.support().iter().all(|c| reverse_project.contains_key(c)))
+                });
+                // Permute arrangements so columns reference what is after the MFP.
+                for key in available_arrangements[index].iter_mut() {
+                    for k in key.iter_mut() {
+                        k.permute_map(&reverse_project);
+                    }
+                }
+                // Currently we only support using arrangements all of whose
+                // key components can be found in some equivalence.
+                // Note: because `order_input` currently only finds arrangements
+                // with exact key matches, the code below can be removed with no
+                // change in behavior, but this is being kept for a future
+                // TODO: expand `order_input`
+                available_arrangements[index].retain(|key| {
+                    key.iter().all(|k| {
+                        let k = input_mapper.map_expr_to_global(k.clone(), index);
+                        equivalences
+                            .iter()
+                            .any(|equivalence| equivalence.contains(&k))
+                    })
+                });
+            }
 
-                // Determine if we can perform delta queries with the existing arrangements.
-                // We could defer the execution if we are sure we know we want one input,
-                // but we could imagine wanting the best from each and then comparing the two.
-                let delta_query_plan = delta_queries::plan(
-                    relation,
-                    &input_mapper,
-                    &available_arrangements,
-                    &unique_keys,
-                    &filters,
-                );
-                let differential_plan = differential::plan(
-                    relation,
-                    &input_mapper,
-                    &available_arrangements,
-                    &unique_keys,
-                    &filters,
-                );
+            let old_implementation = implementation.clone();
+
+            // Determine if we can perform delta queries with the existing arrangements.
+            // We could defer the execution if we are sure we know we want one input,
```

Comment:
Can you re-define `delta_query_plan` and `differential_plan` as lambdas that take a `&mut relation` and call the corresponding plan method:

```rust
let delta_query_plan = |relation: &mut MirRelationExpr| {
    delta_queries::plan(
        relation,
        &input_mapper,
        &available_arrangements,
        &unique_keys,
        &filters,
    )
};
```

You can then change the calls like that:

```rust
*relation = delta_query_plan(relation)
    .or_else(|_| differential_plan(relation))
    .expect("Failed to produce a join plan")
```
so differential planning is only attempted if delta planning fails.

Reply:
Oh, good idea, thanks!

Reply:
Done